discovery/src/main/java/com/ticketmaster/api/discovery/operation/Operation.java

```diff
@@ -0,0 +1,48 @@
+package com.ticketmaster.api.discovery.operation;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import com.ticketmaster.api.discovery.util.Preconditions;
+
+public abstract class Operation<T extends Operation<T>> {
+
+  protected Map<String, String> queryParams = new HashMap<String, String>();
+
+  protected abstract T getThis();
+
+  public Map<String, String> getQueryParameters() {
+    return queryParams;
+  }
+
+  public T locale(String locale) {
+    return withParam("locale", locale);
+  }
+
+  public T withParam(String key, String value) {
+    return withParam(key, value, false);
+  }
+
+  public T withParam(String key, String value, boolean emptyAllowed) {
+    Preconditions.checkNotNull(value);
+    if (!emptyAllowed) {
+      Preconditions.checkArgument(!value.trim().isEmpty());
+    }
+
```

Comment:
Could this be made private and have a public method "withOptionalParam" or similar name public instead? I don't mind boolean flags for simple logic like this but I believe they at least should be hidden.